File: 0045-Jump-Game-II/lc0045.rs
// Rust solution.

impl Solution {
    pub fn jump(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        if n < 2 { return 0; }
        let (mut expect, mut anchor) = (nums[0], nums[0]);
        let mut step = 0;
        for i in 1..n {
            if i > anchor as usize {
                anchor = expect;
                step += 1;
            }
            expect = std::cmp::max(expect, i as i32 + nums[i]);
        }
        step+1
    }
}

File: 0045-Jump-Game-II/lc0045.cc
// C++ solution.

class Solution {
public:
    int jump(vector<int>& nums) {
        int n = nums.size();
        if (n < 2) { return 0; }
        int expect = nums[0], anchor = nums[0], step = 0;
        for (int i = 1; i < n; i++) {
            if (i > anchor) {
                anchor = expect;
                step++;
            }
            expect = max(expect, i+nums[i]);
        }
        return step+1;
    }
};
